### PR TITLE
Install dependencies from GitOps operator

### PR DESCRIPTION
**Pull request by chetan-rns**
The GitOps operator should be capable of configuring and installing GitOps specific dependencies on your OpenShift cluster. This will eliminate the hassle of manual configuration for the argocd/pipelines operator.

Since OLM doesn't support cross-namespace dependency resolution, only the pipeline operator will be installed by OLM. The GitOps operator will manually create namespace (argocd) and install ArgoCD operator by creating operator groups and subscriptions. 

**Workflow:**

1. User will install the GitOps operator from OperatorHub. OLM will install the dependent OpenShift pipeline operator.
2. The reconciler will  install ArgoCD  operator in the `argocd` namespace.
4. The reconciler will then create the CR for the ArgoCD  operator.

Co-authored by: @gaganhegde 
 
**Important Note :** Since the sealed secret operator cannot handle multiple instances, this PR will only install the  ArgoCD and Pipelines Operator as dependencies.

### CODE REVIEW

**Review by wtam2018 — DISMISSED** [review]
LGTM
Please follow up my comment in go.mod to see if it is safe to remove the "replace" statement.

File: deploy/olm-catalog/gitops-operator/manifests/gitops-operator.clusterserviceversion.yaml [hunk 1]
```diff
@@ -2,7 +2,7 @@ apiVersion: operators.coreos.com/v1alpha1
 kind: ClusterServiceVersion
 metadata:
   annotations:
-    containerImage: quay.io/redhat-developer/gitops-backend-operator:v0.0.1
+    containerImage: quay.io/cbanavik/gitops-backend-operator:dep
```
Comment (bigkevmcd) [inline deploy/olm-catalog/gitops-operator/manifests/gitops-operator.clusterserviceversion.yaml]: I don't think this should be in this PR?

File: pkg/apis/pipelines/v1alpha1/gitopsservice_types.go [hunk 2]
```diff
@@ -12,6 +12,7 @@ type GitopsServiceSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
```
Comment (bigkevmcd) [inline pkg/apis/pipelines/v1alpha1/gitopsservice_types.go:13]: Remove this comment, and add a proper comment for the Prefix field, it's embedded in the CRD.

https://github.com/redhat-developer/gitops-operator/pull/10/commits/a27b74582dd4544e860b31eb936a9c197e3c2018#diff-4284bda429d34d9eb75be67605112b07f66752910ea60847e9980bec6b9afccfR35

File: deploy/operator.yaml [hunk 3]
```diff
@@ -16,7 +16,7 @@ spec:
       containers:
         - name: gitops-operator
           # Replace this with the built image name
-          image: REPLACE_IMAGE
+          image: quay.io/cbanavik/gitops-operator:role
```
Comment (bigkevmcd) [inline deploy/operator.yaml]: I think this is automatically replaced, and so we likely don't want your test image.

File: pkg/controller/gitopsservice/dependency/dependency.go [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+		// TODO: Create the operator CR
+	}
+
+	return nil
+}
+
+func (d *Dependency) createResourceIfAbsent(ctx context.Context, obj runtime.Object, ns types.NamespacedName) error {
+	err := d.client.Get(ctx, ns, obj)
+	if err != nil {
+		if errors.IsNotFound(err) {
```
Comment (bigkevmcd) [inline pkg/controller/gitopsservice/dependency/dependency.go:85]: You could probably improve this by using an expression switch and cases for the options (and `default:` returning the error).

**Review by bigkevmcd — CHANGES_REQUESTED** [review]
Reading the image will fail, but that's being addressed, and there are a number of TODOs that it's not clear _when_ they'll be addressed?

File: deploy/olm-catalog/gitops-operator/manifests/gitops-operator.clusterserviceversion.yaml (hunk 1 above)
Comment (chetan-rns) [reply deploy/olm-catalog/gitops-operator/manifests/gitops-operator.clusterserviceversion.yaml]: This is an outdated commit. It was changed earlier

File: deploy/operator.yaml (hunk 3 above)
Comment (chetan-rns) [reply deploy/operator.yaml]: This is outdated as well. I shouldn't have pushed it in the first place :)

File: deploy/crds/pipelines.openshift.io_v1alpha1_gitopsservice_cr.yaml [hunk 5]
```diff
@@ -5,3 +5,4 @@ metadata:
 spec:
   # Add fields here
   size: 3
```
Comment (bigkevmcd) [inline deploy/crds/pipelines.openshift.io_v1alpha1_gitopsservice_cr.yaml:7]: I think we can remove the `size` field, it's not used by the reconciler.

File: cmd/manager/main.go [hunk 6]
```diff
@@ -110,7 +112,7 @@ func main() {
 
 	// Set default manager options
 	options := manager.Options{
-		Namespace:          namespace,
+		Namespace:          "",
 		MetricsBindAddress: fmt.Sprintf("%s:%d", metricsHost, metricsPort),
 	}
```
Comment (wtam2018) [inline cmd/manager/main.go:118]: Can we get rid of the multi namespace support below?

File: pkg/controller/gitopsservice/dependency/dependency_test.go [hunk 7]
```diff
@@ -0,0 +1,179 @@
...
+	"context"
+	"fmt"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+
+	argo "github.com/argoproj-labs/argocd-operator/pkg/apis/argoproj/v1alpha1"
+	v1 "github.com/operator-framework/api/pkg/operators/v1"
+	"github.com/operator-framework/api/pkg/operators/v1alpha1"
```
Comment (wtam2018) [inline pkg/controller/gitopsservice/dependency/dependency_test.go:13]: run `go mod tidy`

File: cmd/manager/main.go (hunk 6 above)
Comment (chetan-rns) [reply cmd/manager/main.go:118]: The GitOps operator only supports the `all-namespace` installation mode and it creates resources across multiple namespaces. If the CR is namespace scoped, the controller would only create resources in the namespace of the CR

Comment (wtam2018) [reply cmd/manager/main.go:118]: I mean can this block of code safely removed?
```
	// Add support for MultiNamespace set in WATCH_NAMESPACE (e.g ns1,ns2)
	// Note that this is not intended to be used for excluding namespaces, this is better done via a Predicate
	// Also note that you may face performance issues when using this with a high number of namespaces.
	// More Info: https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/cache#MultiNamespacedCacheBuilder
	if strings.Contains(namespace, ",") {
		options.Namespace = ""
		options.NewCache = cache.MultiNamespacedCacheBuilder(strings.Split(namespace, ","))
	}
```

File: pkg/controller/gitopsservice/config/config.go [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+	return Config{c}
+}
+
+// ExtractPrefixes will extract the prefixes list from config
+func (c *Config) ExtractPrefixes() []string {
+	prefixes := []string{}
+	if prefixString, ok := c.Data[gitopsPrefix]; ok {
+		prefixes = append(prefixes, strings.Split(prefixString, ",")...)
+	}
+	return prefixes
```
Comment (wtam2018) [inline pkg/controller/gitopsservice/config/config.go:45]:
```suggestion
	prefixes := []string{}
	if prefixString, ok := c.Data[gitopsPrefix]; ok {
		return strings.Split(prefixString, ",")
	}
	return []string{}
```

File: pkg/controller/gitopsservice/dependency/dependency.go [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+}
+
+func addPrefixIfNecessary(prefix, name string) string {
+	if prefix != "" {
+		return prefix + "-" + name
+	}
+	return name
+}
+
+func argoCDCr(ns string) (runtime.Object, string, error) {
```
Comment (wtam2018) [inline pkg/controller/gitopsservice/dependency/dependency.go:209]: function nane should be changed to `argoCDCR`.

File: pkg/controller/gitopsservice/dependency/dependency.go [hunk 10]
```diff
@@ -0,0 +1,269 @@
...
+		Object: map[string]interface{}{
+			"apiVersion": "bitnami.com/v1alpha1",
+			"kind":       "SealedSecretController",
+			"metadata": map[string]interface{}{
+				"name":      name,
+				"namespace": ns,
+			},
+			"spec": map[string]interface{}{
+				"image": map[string]interface{}{
+					"repository": "quay.io/bitnami/sealed-secrets-controller@sha256:8e9a37bb2e1a6f3a8bee949e3af0e9dab0d7dca618f1a63048dc541b5d554985",
```
Comment (wtam2018) [inline pkg/controller/gitopsservice/dependency/dependency.go:251]: We need to be able get image from environment variable in order to support disconnected install.

File: cmd/manager/main.go (hunk 6 above)
Comment (chetan-rns) [reply cmd/manager/main.go:110]: Removed the unwanted code

File: pkg/controller/gitopsservice/dependency/dependency.go (hunk 10 above)
Comment (chetan-rns) [reply pkg/controller/gitopsservice/dependency/dependency.go:257]: Added the support to read images from env variable

**Comment by chetan-rns** [comment]
@bigkevmcd @wtam2018 Removed sealed secret operator dependency. Please consider it for review

**Review by bigkevmcd — APPROVED** [review]

**Pull request merged**